File: conveyor/src/main/java/ru/artemev/conveyor/exception/ErrorHandler.java
package ru.artemev.conveyor.exception;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;

@ControllerAdvice
public class ErrorHandler {
  @ExceptionHandler(value = {BaseException.class})
  public ResponseEntity<ApiError> handlerHandlerException(BaseException ex) {
    ApiError apiError = ex.getApiError();
    HttpStatus httpStatus = ex.getHttpStatus();
    ex.printStackTrace();
    return new ResponseEntity<>(apiError, httpStatus);
  }

  @ExceptionHandler(value = {Exception.class})
  public ResponseEntity<ApiError> defaultHandlerException(Exception ex) {
    ApiError apiError = new ApiError(ex.getClass().getName(), ex.getMessage());
    ex.printStackTrace();
    return new ResponseEntity<>(apiError, HttpStatus.BAD_REQUEST);
  }
}

File: gateway/src/main/kotlin/ru/artemev/gateway/exception/ErrorHandler.kt
package ru.artemev.gateway.exception

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler

@ControllerAdvice
class ErrorHandler {
    @ExceptionHandler(BaseException::class)
    fun handlerHandlerException(ex: BaseException): ResponseEntity<ApiError> {
        ex.printStackTrace()
        return ResponseEntity(ex.apiError, ex.httpStatus)
    }

    @ExceptionHandler(Exception::class)
    fun defaultHandlerException(ex: Exception): ResponseEntity<ApiError> {
        val apiError = ApiError(ex.javaClass.name, ex.message.toString())
        ex.printStackTrace()
        return ResponseEntity(apiError, HttpStatus.BAD_REQUEST)
    }
}
